### PR TITLE
Adjust network

### PR DESCRIPTION
**Pull request by ibab**
Multiple adjustments to the model definition:

1. Use separate 1x1 convolutions for residuals and skipped connections (suggested by @basveeling)
2. We can feed the raw waveform into the first convolutional layer instead of a one-hot encoding (as explained by @jyegerlehner)
3. Made the histogram summaries optional

I'm not super sure about 2 yet, it seems to produce roughly similar results to the one-hot encoding.
And it would probably be better to use quantized values instead of the original signal.

### CODE REVIEW

File: wavenet.py [hunk 1]
```diff
@@ -18,13 +18,15 @@ def __init__(self,
                  dilations,
                  filter_width,
                  residual_channels,
-                 dilation_channels):
+                 dilation_channels,
+                 store_summaries=False):
```
Comment (mecab) [inline wavenet.py]: Should we add a flag to make `store_summaries` be changeable from `train.py`?

**Comment by jyegerlehner** [comment]
Ah, I was working on a PR to do #1, that was branched from the add-biases PR, since I thought you suggested that ordering.

On #2, I tested my version of that code, and also didn't observe that it made any obvious difference, at least as things are now.

File: wavenet.py [hunk 2]
```diff
@@ -53,16 +56,26 @@ def _create_dilation_layer(self, input_batch, layer_index, dilation, in_channels
...
+        # The result of this 1x1 convolution is passed to the next block.
+        weights_residuals = tf.Variable(tf.truncated_normal(
+            [1, dilation_channels, in_channels], stddev=0.2, name="residuals"))
+        residuals = tf.nn.conv1d(out, weights_residuals, stride=1,
+                                 padding="SAME", name="residuals")
 
-        return transformed, input_batch + transformed
+        # The result of this 1x1 convolution is gathered at the end.
+        weights_skipped = tf.Variable(tf.truncated_normal(
+            [1, dilation_channels, in_channels], stddev=0.2, name="skipped"))
```
Comment (jyegerlehner) [inline wavenet.py]: I suggest adding a separate skip_channels param that serves as the number of output channels, instead of "in_channels" AKA residual_channels. I think we want the flexibility to have the skip contribution channels be different than the residual channels.

File: wavenet.py [hunk 3]
```diff
@@ -108,14 +121,15 @@ def _create_network(self, input_batch):
             # Perform (+) -> ReLU -> 1x1 conv -> ReLU -> 1x1 conv to
             # postprocess the output.
             w1 = tf.Variable(tf.truncated_normal(
-                [1, self.residual_channels, int(self.channels / 2)], stddev=0.3,
+                [1, self.residual_channels, self.channels], stddev=0.3,
```
Comment (jyegerlehner) [inline wavenet.py]: I would suggest making the input and output channels here equal to the new skip_channels param, if you went for what I was saying above.

Back on your thread for "generating good audio samples", one of the ideas I mentioned I'd like to try is setting the channels in these last two 1x1/ReLU layers to a larger number, and I think the skip_channels would cleanly fill that role.

File: wavenet.py [hunk 4]
```diff
@@ -108,14 +121,15 @@ def _create_network(self, input_batch):
...
                 name="postprocess1"))
             w2 = tf.Variable(tf.truncated_normal(
-                [1, int(self.channels / 2), self.channels], stddev=0.3,
+                [1, self.channels, self.channels], stddev=0.3,
```
Comment (jyegerlehner) [inline wavenet.py]: And then the input channels here would also be skip_channels.

File: wavenet.py [hunk 5]
```diff
@@ -152,9 +166,10 @@ def predict_proba(self, waveform, name='wavenet'):
 
     def loss(self, input_batch, name='wavenet'):
         with tf.variable_scope(name):
-            input_batch = self._preprocess(input_batch)
-            encoded = self._one_hot(input_batch)
-            raw_output = self._create_network(encoded)
+            quantized = self._preprocess(input_batch)
+            encoded = self._one_hot(quantized)
+            inputs = tf.reshape(input_batch, [self.batch_size, -1, 1])
+            raw_output = self._create_network(inputs)
```
Comment (jyegerlehner) [inline wavenet.py]: That looks familiar. :)

File: wavenet_params.json [hunk 6]
```diff
@@ -2,8 +2,9 @@
     "filter_width": 2,
     "quantization_steps": 256,
     "sample_rate": 16000,
-    "dilations": [1, 2, 4, 8, 16, 32, 64, 128, 256,
-                  1, 2, 4, 8, 16, 32, 64, 128, 256],
+    "dilations": [1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
+                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
+                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512],
     "residual_channels": 32,
     "dilation_channels": 16
```
Comment (jyegerlehner) [inline wavenet_params.json]: "skip_channels": 1024 ? Pretty please?

Comment (jyegerlehner) [inline wavenet_params.json]: Actually, of course it doesn't need to default to anything so big. In consideration for memory requirements.

**Comment by jyegerlehner** [comment]
@ibab When I gave that list of things I'd like to try over on the "Generating good audio samples" I didn't expect you'd go code them all and put them in straight away :) Some of those ideas are a bit wild and thought I'd try them out and bring them back if they worked. I don't mind. Just a bit embarrassed that some of my nutty ideas are being shoe-horned straight into master.

File: wavenet.py (hunk 5 above)
Comment (jyegerlehner) [inline wavenet.py]: Wouldn't corresponding changes also need to be made in generate.py?

**Comment by ibab** [comment]
As 2. hasn't really worked better in my overnight training run, I think I'll just close this PR and we can focus on getting @jyegerlehner's changes merged.

File: wavenet.py (hunk 5 above)
Comment (lemonzi) [inline wavenet.py]: Have we tested this to see if it helps with generation? I'd suggest moving this to a separate PR, as it's quite a big change in the network architecture.

File: wavenet.py [hunk 7]
```diff
@@ -90,13 +103,13 @@ def _create_network(self, input_batch):
         outputs = []
         current_layer = input_batch
 
-        current_layer = self._create_causal_layer(current_layer, self.channels, self.residual_channels)
+        current_layer = self._create_causal_layer(current_layer, 1, self.residual_channels)
```
Comment (lemonzi) [inline wavenet.py]: See L172.

File: wavenet.py (hunk 3 above)
Comment (lemonzi) [inline wavenet.py]: The intermediate tensor could be something else, but the number of hyperparameters is growing out of hand... How about skip_channels is a vector that defines the channels per layer, with optionally deeper convolutions than the 2 layers we have now? See https://github.com/ibab/tensorflow-wavenet/pull/14/files#diff-59f2c602fd3753c87268cb878555ae38R135 for an implementation.

**Comment by lemonzi** [comment]
Oh, didn't see that, my bad. @jyegerlehner, what do you think about this skip_channels vector?

**Comment by jyegerlehner** [comment]
@lemonzi  I like the idea in principle but am am bit torn as it is more complicated. Let's see if we can get opinions from ibab and others. Do you think it needs to be included in [PR 62](https://github.com/ibab/tensorflow-wavenet/pull/62) or would you prefer to bring it in as your own separate PR?

**Comment by ibab** [comment]
@jyegerlehner @lemonzi: `skip_channels` is a good idea. Let's bring it in as a separate PR to make things easier to review.

I prefer `add_summaries` as well, after the reminder that it's kind of a TensorFlow convention.
I can extract the optional summaries part into a separate PR.
